Avoid error when trying to translate an invalid module

PululuK opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Check to avoid the error: : ```Fatal Error: Call to a member function isUsingNewTranslationSystem() on null```  when translating your modules ! This can happen if you have modules that have been removed, installed, or disabled incorrectly..
| Type?         | bug fix 
| Category?     |  CO 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | N/A
| How to test?  | https://drive.google.com/file/d/1-QJ-4eJg1BiiNuSddXZ0U4AmvBC3jAJ4/view

### How to reproduce ?
https://drive.google.com/file/d/1-QJ-4eJg1BiiNuSddXZ0U4AmvBC3jAJ4/view

### How to test the PR ?
Check that you no longer have the fatal error, but a cleaner error message indicating that the module concerned is invalid

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/15871)
<!-- Reviewable:end -->

PululuK committed:
Modules new translation system

Check to avoid the error: : ```Fatal Error: Call to a member function isUsingNewTranslationSystem() on null```  when translating your modules ! This can happen if you have modules that have been removed, installed, or disabled incorrectly.

PululuK committed:
Modules new translation system

Check to avoid the error: : ```Fatal Error: Call to a member function isUsingNewTranslationSystem() on null```  when translating your modules ! This can happen if you have modules that have been removed, installed, or disabled incorrectly.

PululuK committed:
Modules new translation system

Check to avoid the error: : ```Fatal Error: Call to a member function isUsingNewTranslationSystem() on null```  when translating your modules ! This can happen if you have modules that have been removed, installed, or disabled incorrectly.

Progi1984 commented:
@PululuK Hi, thank you for your contribution. Please could you give us a scenario for testing this PR ? Thanks

matthieu-rolland reviewed (dismissed):
EDIT: actually, before validating I would like the opinion of another core dev

@PierreRambaud @jolelievre @Progi1984:

What do you think should happen in this method if the module is absent because it was badly removed (or for another weird reason). Do you guys think it should throw an exception ?

PierreRambaud commented:
Seems legit for me :+1:

PierreRambaud commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -203,7 +204,10 @@ public function findRouteParameters(ParameterBag $query)
-        return $module->isUsingNewTranslationSystem();
+        if (Validate::isLoadedObject($module)) {
+            return $module->isUsingNewTranslationSystem();
+        } else {
```
```suggestion
        }
```

PierreRambaud commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -203,7 +204,10 @@ public function findRouteParameters(ParameterBag $query)
+        if (Validate::isLoadedObject($module)) {
+            return $module->isUsingNewTranslationSystem();
+        } else {
+            return false;
```
```suggestion

        return false;
```

PierreRambaud commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -203,7 +204,10 @@ public function findRouteParameters(ParameterBag $query)
+            return $module->isUsingNewTranslationSystem();
+        } else {
+            return false;
+        }
```
```suggestion
```

PululuK committed:
Update src/Adapter/Translations/TranslationRouteFinder.php

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

PululuK committed:
Update src/Adapter/Translations/TranslationRouteFinder.php

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

PululuK committed:
Update src/Adapter/Translations/TranslationRouteFinder.php

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

Robin-Fischer-PS commented:
Hi @PululuK , thanks for the PR. Could you give us a scenario for testing this PR ? Thanks ;)

PululuK commented:
> Hi @PululuK , thanks for the PR. Could you give us a scenario for testing this PR ? Thanks ;)

Hello @Robin-Fischer-PS  @PierreRambaud ,

We had serious problems with a site, which I unfortunately can not expose here!
I will give a very basic example to reproduce this based on the principle **Do not trust user input**
 :

As you can see, the module name is retrieved from the URL parameters after the form is submitted via the GEST method. So we can send anything.

Thanks for taking a look here please?

https://drive.google.com/file/d/1-QJ-4eJg1BiiNuSddXZ0U4AmvBC3jAJ4/view

eternoendless commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -203,7 +204,10 @@ public function findRouteParameters(ParameterBag $query)
+        }
 
-        return $module->isUsingNewTranslationSystem();
+        return false;
```
This should throw an Exception if the module is not valid, not silently return false.

There should also be a try/catch block in TranslationsController::modifyTranslationsAction with a flash error message and a redirection to the overview page.

PululuK replied:
Hello @eternoendless Thanks !

The legacy controller is still present for modules not used by the new translation system. We agree that `AdminTranslationsController` (legacy) will also be impacted?

I do not think that modifs only at this level `TranslationsController :: modifyTranslationsAction` can cover all impacted zones.

What do you think ?

eternoendless commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -31,6 +31,7 @@
 use PrestaShopBundle\Service\TranslationService;
 use Symfony\Component\HttpFoundation\ParameterBag;
 use Symfony\Component\PropertyAccess\PropertyAccess;
+use Validate;
```
```suggestion
use Module;
```

eternoendless commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -203,7 +204,10 @@ public function findRouteParameters(ParameterBag $query)
     private function isModuleUsingNewTranslationSystem($moduleName)
     {
         $module = $this->moduleRepository->getInstanceByName($moduleName);
+        if (Validate::isLoadedObject($module)) {
```
I think this check would be more adequate:

```suggestion
        if ($module instanceof Module) {
```

eternoendless reviewed (changes requested):
Hi @PululuK 

Thanks for this contribution! Some adjustments need to be done for this use case to be properly managed.

PululuK committed:
Update src/Adapter/Translations/TranslationRouteFinder.php

Co-Authored-By: Pablo Borowicz <pablo.borowicz@prestashop.com>

PululuK committed:
Update src/Adapter/Translations/TranslationRouteFinder.php

Co-Authored-By: Pablo Borowicz <pablo.borowicz@prestashop.com>

PululuK committed:
Trigger an exception when the module is invalid

PululuK committed:
Add translator in TranslationRouteFinder

PululuK committed:
Trigger an exception when the module is invalid

PululuK commented:
Hi @PierreRambaud @eternoendless @matthieu-rolland , patches have been applied.
Can you check please ? Thanks !

PierreRambaud commented:
@PululuK PrettyCi needs to be run :)

PululuK committed:
Modules new translation system

Trigger an exception when the module is invalid and Add translator in TranslationRouteFinder

PululuK commented:
Hi @PierreRambaud 

Should I do something?

Thank you

PululuK commented:
> @PululuK PrettyCi needs to be run :)

Hi @PierreRambaud

Should I do something?

Thank you

PierreRambaud commented:
You can check https://github.com/PrestaShop/PrestaShop/pull/15871/checks?check_run_id=287417626
You have unwanted spaces :)

PululuK committed:
Removing unwanted spaces

PululuK commented:
>  You have unwanted spaces :)

This is not the first time that **PrettyCI** has been trying to show me her muscles :)
Thank you @PierreRambaud

eternoendless commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -72,19 +75,27 @@ class TranslationRouteFinder
         Link $link,
-        ModuleRepositoryInterface $moduleRepository
+        ModuleRepositoryInterface $moduleRepository,
+        TranslatorInterface $translator
```
I'm not sure about injecting the translator here. In addition, it would be considered a breaking change.

PululuK replied:
I completely agree ! 
Why make it complicated when it can be simple ? 😀

Done

eternoendless commented on src/Adapter/Translations/TranslationRouteFinder.php:
```diff
@@ -204,6 +215,10 @@ private function isModuleUsingNewTranslationSystem($moduleName)
         $module = $this->moduleRepository->getInstanceByName($moduleName);
 
+        if (!($module instanceof Module)) {
+            throw new Exception($this->translator->trans('Invalid module : %s', array($moduleName), 'Admin.International.Notification'));
```
I think it would be best to throw an untranslated `PrestaShopBundle\Exception\InvalidModuleException`, which would have to be catched in `TranslationsController::modifyTranslationsAction()` and transformed into a proper error message.

For this you'd need to [add a flash message](https://symfony.com/doc/3.4/controller.html#flash-messages) and redirect back to the `admin_international_translations_show_settings` route.

And then, I think a better error message would be: "An error has occurred, this module does not exist: %s" which already already exists in `Admin.International.Notification`

PululuK replied:
Thank you @eternoendless  ! Done

PululuK committed:
Avoid error when trying to translate an invalid module

PululuK committed:
Fix PrettyCI

PululuK committed:
Fix PrettyCI

PululuK commented:
Hi @LouiseBonnard @khouloudbelguith 
Can you remove **waiting for author** label please ?
Thank you

marionf commented:
Guys, what is the status on this PR ? Does it needs changes ? Or is it ready for QA ?

eternoendless commented:
This PR needs to be rebased into 1.7.7.x

PululuK commented:
> This PR needs to be rebased into 1.7.7.x

Can you help me to rebase please ?

eternoendless commented:
I did it myself, don't worry @PululuK :)

eternoendless commented:
By the way, the video linked in the "how to test" section is no longer available. Would you please re-upload it?

PululuK commented:
> By the way, the video linked in the "how to test" section is no longer available. Would you please re-upload it?

Done !

SD1982 commented:
LGTM Thanks @PululuK

PierreRambaud commented:
Thanks @PululuK
